src/transformers/modeling_utils.py

```diff
@@ -4000,9 +4005,24 @@ def _fix_key(key):
                     _loaded_keys = [k[len(prefix) + 1 :] for k in loaded_keys]
                 else:
                     _loaded_keys = loaded_keys
-                set_initialized_submodules(model, _loaded_keys)
+                not_initialized_submodules = set_initialized_submodules(model, _loaded_keys)
+            else:
+                not_initialized_submodules = dict(model.named_modules())
             # This will only initialize submodules that are not marked as initialized by the line above.
-            model.apply(model._initialize_weights)
+            if is_deepspeed_zero3_enabled():
+                import deepspeed
+
+                not_initialized_parameters = list(
+                    set(
+                        itertools.chain.from_iterable(
+                            submodule.parameters(recurse=False) for submodule in not_initialized_submodules.values()
+                        )
+                    )
+                )
+                with deepspeed.zero.GatheredParameters(not_initialized_parameters, modifier_rank=0):
```

Comment:
Nice! This is efficient when compared to the previous commit wherein it was gathering every module in the `_initialize_weights`. Now, only the `not_initialized_parameters` are gathered to be properly initialized via the correct `torch.nn.init` method while the rest of the modules are loaded from the checkpoint.